### PR TITLE
Pass identity information into RpcMethod from Http Service

### PR DESCRIPTION
**Pull request by rain-on**
To support Privacy in a multi-tenancy operating mode, privacy-oriented requests are required to be submitted with a JWT containing the Enclave public key to be used for the operation.

This enclave key is _not_ part of the Json RPC Body, but rather part of the header - thus the JsonRpcRequest class has been renamed JsonRpcRequestBody (and represents _only_ the body of the request), and a new JsonRpcRequest class has been created (comprising a User, and also the a JsonRpcRequestBody).

All tests, and methods have been updated to handle the additional layer of indirection.

Signed-off-by: Trent Mohay <trent.mohay@consensys.net>

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java [hunk 1]
```diff
@@ -14,94 +14,39 @@
...
-    return version;
+  public JsonRpcRequest(final JsonRpcRequestBody requestBody) {
+    this(requestBody, Optional.empty());
   }
 
-  @JsonInclude(Include.NON_NULL)
-  @JsonGetter("params")
-  public Object[] getParams() {
-    return params;
+  public JsonRpcRequest(final JsonRpcRequestBody requestBody, final Optional<User> user) {
```
Comment (jframe) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java]: The user doesn't need to be optional as there another constructor to allow for the none case

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java [hunk 2]
```diff
@@ -14,94 +14,39 @@
...
-  public boolean isNotification() {
-    return isNotification;
+  public JsonRpcRequestBody getRequestBody() {
+    return requestBody;
   }
 
-  @JsonIgnore
-  public int getParamLength() {
-    return hasParams() ? params.length : 0;
+  public Optional<User> getIdentity() {
```
Comment (jframe) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java]: What's the distinction between user and identity? Perhaps this should be just getUser?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketRequestHandler.java [hunk 3]
```diff
@@ -76,7 +77,8 @@ public void handle(
             LOG.debug("WS-RPC request -> {}", request.getMethod());
             request.setConnectionId(id);
             if (AuthenticationUtils.isPermitted(authenticationService, user, method)) {
-              future.complete(method.response(request));
+              final JsonRpcRequest fullRequest = new JsonRpcRequest(request, user);
```
Comment (jframe) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketRequestHandler.java]: What's the distinction between request and fullRequest?

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapper.java [hunk 4]
```diff
@@ -85,8 +85,8 @@ public UnsubscribeRequest mapUnsubscribeRequest(final JsonRpcRequest jsonRpcRequ
   }
 
   private WebSocketRpcRequest validateRequest(final JsonRpcRequest jsonRpcRequest) {
-    if (jsonRpcRequest instanceof WebSocketRpcRequest) {
-      return (WebSocketRpcRequest) jsonRpcRequest;
+    if (jsonRpcRequest.getRequestBody() instanceof WebSocketRpcRequest) {
```
Comment (jframe) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapper.java]: This reads a bit odd. Thinking should rename WebSocketRpcRequest to WebSocketRpcRequestBody as well

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthSubscribeIntegrationTest.java [hunk 5]
```diff
@@ -62,7 +62,7 @@ public void before() {
   public void shouldAddConnectionToMap(final TestContext context) {
     final Async async = context.async();
 
-    final JsonRpcRequest subscribeRequest = createEthSubscribeRequest(CONNECTION_ID_1);
+    final JsonRpcRequestBody subscribeRequest = createEthSubscribeRequest(CONNECTION_ID_1);
```
Comment (jframe) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthSubscribeIntegrationTest.java]: rename createEthSubscribeRequest to createEthSubscribeRequestBody

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthUnsubscribeIntegrationTest.java [hunk 6]
```diff
@@ -63,7 +63,7 @@ public void shouldRemoveConnectionWithSingleSubscriptionFromMap(final TestContex
     final Long subscriptionId = subscriptionManager.subscribe(subscribeRequest);
     assertThat(subscriptionManager.getSubscriptionById(subscriptionId)).isNotNull();
 
-    final JsonRpcRequest unsubscribeRequest =
+    final JsonRpcRequestBody unsubscribeRequest =
```
Comment (jframe) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthUnsubscribeIntegrationTest.java]: rename to unsubscribeRequestBody

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthUnsubscribeIntegrationTest.java [hunk 7]
```diff
@@ -95,7 +95,7 @@ public void shouldRemoveSubscriptionAndKeepConnection(final TestContext context)
     assertThat(subscriptionManager.getSubscriptionById(subscriptionId1)).isNotNull();
     assertThat(subscriptionManager.getSubscriptionById(subscriptionId2)).isNotNull();
 
-    final JsonRpcRequest unsubscribeRequest =
+    final JsonRpcRequestBody unsubscribeRequest =
```
Comment (jframe) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthUnsubscribeIntegrationTest.java]: same as above. rename to unsubscribeRequestBody

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapperTest.java [hunk 8]
```diff
@@ -54,83 +55,88 @@ public void before() {
 
   @Test
   public void mapRequestToUnsubscribeRequest() {
-    final JsonRpcRequest jsonRpcRequest =
+    final JsonRpcRequestBody jsonRpcRequest =
```
Comment (jframe) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapperTest.java]: nit: these request variable names should be jsonRpcRequestBody instead

File: ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1EthProxyProtocol.java [hunk 9]
```diff
@@ -93,9 +94,9 @@ public void onClose(final StratumConnection conn) {}
   public void handle(final StratumConnection conn, final String message) {
     try {
       JsonRpcRequest req = new JsonObject(message).mapTo(JsonRpcRequest.class);
-      if ("eth_getWork".equals(req.getMethod())) {
-        sendNewWork(conn, req.getId());
-      } else if ("eth_submitWork".equals(req.getMethod())) {
+      if ("eth_getWork".equals(req.getRequestBody().getMethod())) {
```
Comment (jframe) [inline ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1EthProxyProtocol.java]: nit: extract req.getRequestBody().getMethod() to local var

File: ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1Protocol.java [hunk 10]
```diff
@@ -83,9 +83,9 @@ public Stratum1Protocol(final String extranonce) {
 
   @Override
   public boolean canHandle(final String initialMessage, final StratumConnection conn) {
-    JsonRpcRequest req;
+    JsonRpcRequestBody req;
```
Comment (jframe) [inline ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1Protocol.java]: rename req to reqBody

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java [hunk 11]
```diff
@@ -14,98 +14,39 @@
...
-    return params;
+  public JsonRpcRequest(final JsonRpcRequestBody requestBody, final Optional<User> user) {
+    this.requestBody = requestBody;
+    this.user = user;
   }
 
-  @JsonIgnore
-  public boolean isNotification() {
-    return isNotification;
+  public JsonRpcRequestBody getRequestBody() {
```
Comment (CjHare) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java]: I'd suggest keeping the `requestBody` encapsulated and proving the request through to the `requestBody` (i.e mediator pattern)

```
public Object getId(){
  return requestBody.getId();
}
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java (hunk 1 above)
Comment (rain-on) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java]: done.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java (hunk 2 above)
Comment (rain-on) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java]: done - was written before finding the User class 👍

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketRequestHandler.java (hunk 3 above)
Comment (rain-on) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/WebSocketRequestHandler.java]: FullRequest is now "requestContext" - and contains the request, and the user

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapper.java (hunk 4 above)
Comment (rain-on) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapper.java]: done.

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthSubscribeIntegrationTest.java (hunk 5 above)
Comment (rain-on) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthSubscribeIntegrationTest.java]: done

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthUnsubscribeIntegrationTest.java (hunk 6 above)
Comment (rain-on) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/methods/EthUnsubscribeIntegrationTest.java]: done

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapperTest.java (hunk 8 above)
Comment (rain-on) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/websocket/subscription/request/SubscriptionRequestMapperTest.java]: done

File: ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1EthProxyProtocol.java (hunk 9 above)
Comment (rain-on) [reply ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1EthProxyProtocol.java]: think there's something wrong here - have changed this to use a JsonRequestBody ... the Json Deserialise was never going to work ... :/

File: ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1Protocol.java (hunk 10 above)
Comment (rain-on) [reply ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/Stratum1Protocol.java]: done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java (hunk 11 above)
Comment (rain-on) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java]: move JsonRpcRequest to JsonRpcRequestContext, making this irrelevant (I think?) - ok?

Comment (CjHare) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/JsonRpcRequest.java]: 👍

**Pull request merged**